### PR TITLE
[FEATURE][needs-docs] Add parameters to 'Execute SQL' algorithm

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

Hi,

This PR adds a way to write SQL queries with expressions in the `Execute SQL` algorithm.

It's also possible to use variables coming from the modeler:

executesql3

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nyalldawson** [comment]
Hey @pblottiere 
Can you elaborate a bit on the use case for the parameters? (vs just writing them directly in the query) I can understand the value of expressions but am unsure when the parameters would be valuable (unless they are exposed for use in models via https://github.com/qgis/QGIS/pull/6504)

File: python/plugins/processing/algs/qgis/ui/ExecuteSQLWidget.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        scope = QgsExpressionContextScope()
+
+        for i in range(self.mTreeParameters.topLevelItemCount()):
+            item = self.mTreeParameters.topLevelItem(i)
+            if not item:
+                continue
+
+            scope.setVariable(item.text(0), item.text(1), True)
+
+        return scope
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/ui/ExecuteSQLWidget.py]: Parameter values must be QVariant compatible - I'd suggest leaving this as a dict and converting it to a scope only when the algorithm is executed

File: python/plugins/processing/algs/qgis/ExecuteSQL.py [hunk 2]
```diff
@@ -112,21 +133,30 @@ def processAlgorithm(self, parameters, context, feedback):
...
+
         df = QgsVirtualLayerDefinition()
         for layerIdx, layer in enumerate(layers):
             df.addSource('input{}'.format(layerIdx + 1), layer.id())
 
         if query == '':
             raise QgsProcessingException(
                 self.tr('Empty SQL. Please enter valid SQL expression and try again.'))
+        elif scope:
+            localContext = QgsExpressionContext()
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/ExecuteSQL.py]: Here you should use the full processing algorithm expression context as a start:

    localContext  = self.createExpressionContext( parameters, context )
 
Then the expressions can utilise the global/project/algorithm scopes too.

**Comment by haubourg** [comment]
> Hey @pblottiere

> Can you elaborate a bit on the use case for the parameters? (vs just writing them directly in the query) I can understand the value of expressions but am unsure when the parameters would be valuable (unless they are exposed for use in models via #6504)
> @nyalldawson 

Hi @nyalldawson. 
Thanks for raising that point, the #6504 opens up the possibilities to having parameters in the modeler scope which is really great. I'm asking our customer if that new situation changes the expression of their need. and come back here. 
ping @mhugo, if you have insights here, let us know.

**Comment by haubourg** [comment]
tested and approved for a user point of view. 
See https://vimeo.com/264601587 for a live demo (but with french comments, be warned)

**Comment by pblottiere** [comment]
The `ExecuteSQL` algorithm may be used with parameters coming from the modeler now.

> tested and approved for a user point of view.

Thank you @haubourg, I'll merge this PR tomorrow if I have no more comments in the meantime.

**Comment by nyalldawson** [comment]
Hey @pblottiere what's the difference between expression sources and expressionText sources?

**Comment by pblottiere** [comment]
Hi @nyalldawson,

> what's the difference between expression sources and expressionText sources?

I took that idea from the `QgsExpression::replaceExpressionText()` static method and default Python actions. For example for `Clicked coordinates` action:

``` python
from qgis.PyQt import QtWidgets

QtWidgets.QMessageBox.information(None, "Clicked coords", "layer: [% @layer_id %]\ncoords: ([% @click_x %],[% @click_y %])")
```

In my case, I have a SQL query which may use several expressions, depending on several variables:

```
select * from [% @table %] where Region = '[% @region %]'
``` 

So I considered a new `expressionText` source with multiple 'sub-expression' to replace.

But maybe you have a better idea to manage such cases?

Thanks.

**Comment by pblottiere** [comment]
Hi @nyalldawson!

> @pblottiere I'll re-review tomorrow

Have you forgotten me? :)

**Comment by mhugo** [comment]
I've merged. @nyalldawson do not hesitate if you have comments, @pblottiere will still be there to fix :)

**Pull request merged**